### PR TITLE
HIVE-24403: change min_history_level to make it compatible with older versions

### PR DESCRIPTION
**Pull request by pvargacl**
### What changes were proposed in this pull request?
Modify HIVE-23107 to make it backward compatible with older version HMS-s using the same backend DB

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit tests

### CODE REVIEW

**Comment by pvargacl** [comment]
Initial commit to see test results. Missing tests and filtering missing table exception.

**Comment by pvargacl** [comment]
@pvary can I ask you for a review?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java [hunk 1]
```diff
@@ -98,16 +98,24 @@ public void run() {
           handle = txnHandler.getMutexAPI().acquireLock(TxnStore.MUTEX_KEY.Cleaner.name());
           startedAt = System.currentTimeMillis();
           long minOpenTxnId = txnHandler.findMinOpenTxnIdForCleaner();
+          long minTxnIdSeenOpen = txnHandler.findMinTxnIdSeenOpen();
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:101]: could we skip this extra db call if we have `metastore.txn.use.minhistorylevel=false`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java [hunk 2]
```diff
@@ -232,15 +232,23 @@ public static void cleanDb(Configuration conf) throws Exception {
...
-          LOG.error("Error initializing sequence values", e);
-          success = false;
+        String dbProduct = conn.getMetaData().getDatabaseProductName();
+        DatabaseProduct databaseProduct = determineDatabaseProduct(dbProduct, conf);
+        try {
+          resetTxnSequence(databaseProduct, stmt);
+          stmt.executeUpdate("INSERT INTO \"NEXT_LOCK_ID\" VALUES(1)");
+          stmt.executeUpdate("INSERT INTO \"NEXT_COMPACTION_QUEUE_ID\" VALUES(1)");
+        } catch (SQLException e) {
+          if (!databaseProduct.isTableNotExists(e)) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java:244]: Previous version was much more readable and concise.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java [hunk 3]
```diff
@@ -385,6 +391,26 @@ public static String queryToString(Configuration conf, String query, boolean inc
...
   }
 
+  /**
+   * This is only for testing, it does not use the connectionPool from TxnHandler!
+   * @param conf
+   * @param query
+   * @throws Exception
+   */
+  @VisibleForTesting
+  public static void executeUpdate(Configuration conf, String query)
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java:401]: That's not a test class. Production code becomes massive because of that.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 4]
```diff
@@ -390,6 +404,42 @@ public void setConf(Configuration conf){
...
+      // don't check it if disabled
+      return false;
+    }
+    Connection dbConn = null;
+    boolean tableExists = true;
+    try {
+      dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
+      try (Statement stmt = dbConn.createStatement()) {
+        // Dummy query to see if table exists
+        try (ResultSet rs = stmt.executeQuery("SELECT MIN(\"MHL_MIN_OPEN_TXNID\") FROM \"MIN_HISTORY_LEVEL\"")) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:423]: you can just select 1

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 5]
```diff
@@ -670,6 +725,8 @@ public OpenTxnsResponse openTxns(OpenTxnRequest rqst) throws MetaException {
 
       assert txnIds.size() == numTxns;
 
+      addTxnToMinHistoryLevel(dbConn, txnIds, minOpenTxnId);
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:728]: why not to put this under above useMinHistoryLevel check?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 6]
```diff
@@ -5094,6 +5153,99 @@ public void countOpenTxns() throws MetaException {
...
+      for (String insert : inserts) {
+        LOG.debug("Going to execute insert <" + insert + ">");
+        stmt.execute(insert);
+      }
+      LOG.info("Added entries to MIN_HISTORY_LEVEL for current txns: (" + txnIds + ") with min_open_txn: " + minOpenTxnId);
+    } catch (SQLException e) {
+      if (dbProduct.isTableNotExists(e)) {
+        // If the table does not exists anymore, we disable the flag and start to work the new way
+        // This enables to switch to the new functionality without a restart
+        useMinHistoryLevel = false;
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5185]: Why is this needed? Are you covering the case that schema change is done while old HMS is still running?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 7]
```diff
@@ -1166,6 +1166,55 @@ public long findMinOpenTxnIdForCleaner() throws MetaException {
...
+    }
+    Connection dbConn = null;
+    try {
+      try {
+        dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
+        long minOpenTxn;
+        try (Statement stmt = dbConn.createStatement()) {
+          try (ResultSet rs = stmt.executeQuery("SELECT MIN(\"MHL_MIN_OPEN_TXNID\") FROM \"MIN_HISTORY_LEVEL\"")) {
+            if (!rs.next()) {
+              throw new IllegalStateException("Scalar query returned no rows?!");
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1190]: is this even possible?

**Review by deniskuzZ — COMMENTED** [review]
LGTM, minor comments

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java [hunk 8]
```diff
@@ -186,6 +186,19 @@ public boolean isDeadlock(SQLException e) {
             || e.getMessage().contains("can't serialize access for this transaction"))));
   }
 
+  /**
+   * Is the given exception a table not found exception
+   * @param e Exception
+   * @return
+   */
+  public boolean isTableNotExists(SQLException e) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java:194]: maybe rename to `isTableNotExistsError`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java (hunk 3 above)
Comment (pvargacl) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java:401]: Well, this class is a test utility.
/**
 * Utility methods for creating and destroying txn database/schema, plus methods for
 * querying against metastore tables.
 * Placed here in a separate class so it can be shared across unit tests.
 */
public final class TxnDbUtil

The problem is more like getEpochFn and executeQueriesInBatchNoCount was added to this class, those are production code. I know it would be nicer if it would be in a test package, but then it would be harder to use in 5 different projects

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 4 above)
Comment (pvargacl) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:424]: fixed

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 5 above)
Comment (pvargacl) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:729]: That was my first intent, but it resulted in lock timeout, after inserting the new records in the txns table, the min open select was not running

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java (hunk 3 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java:401]: in this case we should consider refactoring this class

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 6 above)
Comment (pvargacl) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5198]: The idea is multiple HMS is using the same backend db, you upgrade them one by one, the last one changes the schema, all the others change to the new functionality, after the first call to min_history table. Do you have any practical example of wrapping them in aspect, I do not want to much more code complexity just to avoid checking an exception in four places.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java (hunk 8 above)
Comment (pvargacl) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java:194]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java (hunk 3 above)
Comment (pvargacl) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnDbUtil.java:401]: Created a Jira for that: https://issues.apache.org/jira/browse/HIVE-24477

**Review by deniskuzZ — APPROVED** [review]
LGTM, pending tests
